clients/drcachesim/tracer/tracer.h

```diff
@@ -0,0 +1,239 @@
+     * the same user_data path that drmgr does.
+     */
+    bool repstr;
+    bool scatter_gather;
+#ifdef HAS_SNAPPY
+    snappy_file_writer_t *snappy_writer;
+#endif
+#ifdef HAS_ZLIB
+    z_stream zstream;
+    byte *buf_compressed;
+#endif
+#ifdef HAS_LZ4
+    LZ4F_compressionContext_t lzcxt;
+    size_t buf_lz4_size;
+    byte *buf_lz4;
+#endif
+    bool has_thread_header;
+    // The physaddr_t class is designed to be per-thread.
+    physaddr_t physaddr;
+    uint64 num_phys_markers;
+    byte *v2p_buf;
+    uint64 num_v2p_writeouts; /* v2p_buf writeout instances. */
+} per_thread_t;
+
+/* Allocated TLS slot offsets */
+enum {
+    MEMTRACE_TLS_OFFS_BUF_PTR,
+    /* XXX: we could make these dynamic to save slots when there's no
+     *-L0I_filter or -L0D_filter.
+     */
+    MEMTRACE_TLS_OFFS_DCACHE,
+    MEMTRACE_TLS_OFFS_ICACHE,
+    /* The instruction count for -L0I_filter. */
+    MEMTRACE_TLS_OFFS_ICOUNT,
+    /* The decrementing instruction count for -trace_after_instrs.
+     * We could share with MEMTRACE_TLS_OFFS_ICOUNT if we cleared in each thread
+     * on the transition.
+     */
+    MEMTRACE_TLS_OFFS_ICOUNTDOWN,
+    // For has_tracing_windows(), this is the ordinal of the tracing window at
+    // the start of the current trace buffer.  It is -1 if windows are not enabled.
+    MEMTRACE_TLS_OFFS_WINDOW,
+    MEMTRACE_TLS_COUNT, /* total number of TLS slots allocated */
+};
+
+extern reg_id_t tls_seg;
+extern uint tls_offs;
+extern int tls_idx;
+#define TLS_SLOT(tls_base, enum_val) \
+    (((void **)((byte *)(tls_base) + tls_offs)) + (enum_val))
+#define BUF_PTR(tls_base) *(byte **)TLS_SLOT(tls_base, MEMTRACE_TLS_OFFS_BUF_PTR)
+
+extern instru_t *instru;
+
+/* Lock for instrumentation phase transitions and global counters. */
+extern void *mutex;
+
+bool
+is_first_nonlabel(void *drcontext, instr_t *instr);
+
+extern std::atomic<ptr_int_t> tracing_disabled;
+extern std::atomic<ptr_int_t> tracing_window;
+
+/* We have two modes: count instructions only, and full trace.
+ *
+ * XXX i#3995: To implement -max_trace_size with drbbdup cases (with thread-private
+ * encodings), or support nudges enabling tracing, or have a single -trace_for_instrs
+ * transition to something lower-cost than counting, we will likely add a 3rd mode that
+ * has zero instrumentation.  We also would use the 3rd mode for just -trace_for_instrs
+ * with no -retrace_every_instrs.  For now we have just 2 as the case dispatch is more
+ * efficient that way.
+ */
+enum {
+    BBDUP_MODE_TRACE = 0,
+    BBDUP_MODE_COUNT = 1,
+};
+
+#if defined(X86_64) || defined(AARCH64)
+#    define DELAYED_CHECK_INLINED 1
+#else
+/* XXX we don't have the inlining implemented yet for 32-bit architectures. */
+#endif
+
+#define INSTR_COUNT_LOCAL_UNIT 10000
+
+/***************************************************************************
+ * Buffer writing to disk.
+ */
+
+// NOCHECK move to output.h?
```

Comment:
This was supposed to be addressed.

Reply:
Removed.  Ended up deciding to leave the content where it is.